### ISSUE TITLE
Allow using custom Passport LocalStrategy for local authentication

### ISSUE THREAD

**Issue by kelleg1** [iaincollins/next-auth#18]
This is related to https://github.com/iaincollins/next-auth/issues/9 as it is about doing local authentication, but it is about using a custom Passport LocalStrategy I have written. Is it possible to use Passport LocalStrategy with next-auth? I tried creating a provider and passing the strategy in the Strategy property, but that wouldn't work. I wonder if I am doing something wrong. Does it only work with OAuth providers?

```
providers.push({
  providerName: 'Local Login',
  providerOptions: {
    scope: ['profile', 'email'],
  },
  Strategy: require('./passport/local-login'),
  strategyOptions: {
  },
  getProfile(profile) {
    // Normalize profile into one with {id, name, email} keys
    return {
      id: profile.id,
      name: profile.displayName,
      email: profile.emails[0].value,
    };
  }
});
```

**Comment by iaincollins** [comment]
The current system is only designed to support oAuth explicitly and doesn't provide support for LocalStrategies, although it's a reasonable request I'd be happy to accommodate in future.

For now, if you just want to support logging in with a username and password you can add a route that calls `req.login()` and does something like this:

````javascript
expressApp.get(`/auth/password`, (req, res) => {
  UserCollection.find({ email: req.body.email })
  .then(user => {
    if (user && user.password = encryptPassword(req.body.password)))
      req.login(user, (err) => {
        return res.redirect(`/auth/callback?action=signin&service=password`)
      })
    } else {
      return res.redirect(`/auth/error?action=signin&type=password-failure`)
    }
  })
  .catch(err => {
    return res.redirect(`/auth/error?action=signin&type=password-error`)
  })
})
````

If you'd like to see an example of an `index.js` that that invokes NextAuth and sets up custom routes after invoking it [take a look at this example](https://github.com/iaincollins/nextjs-starter/blob/master/index.js).

Note: Be sure not to pass a 'port' to NextAuth if you want to add custom routes. As in the example you will have to define default routing for Next.js and call expressApp.listen() after you have finished declaring any custom routings.

For some background, this is what the email token sign in feature does and will give you exactly the same behaviour (and so you will be able to also support linking with oAuth accounts too, if you like).

I am happy to look at making this easier in future through direct support for localStrategies - and also looking at adding the ability to turn off email only sign in, perhaps in favour of some sort of user registration option, if that would make more sense for some people.

Generally email token based sign in (as used by sites like Slack and Medium) is regarded as good practice from a security perspective - even if the implementation in NextAuth is currently a bit simple (the tokens are single use, but don't expire) - as it prevents the possibility that even encrypted passwords might be leaked (because they are all transitory and single use).

So, I'm included to encourage to people to use that option by default - but I appreciate it's not always suitable for all use cases and am happy to look at supporting other scenarios, including things like localStrategies that 2FA.

**Comment by kelleg1** [comment]
@iaincollins I did this, added a password field to the signin component, and updated my form action URL to /auth/password and it works with JavaScript disabled, but when JavaScript is enabled the call to NextAuth.signin() in handleSubmit() ends up sending the verification email anyway, I think because it does the POST to /auth/email/signin. Should I not call NextAuth.signin()? Can I use next-auth/next-auth client when wanting to have password authentication?

**Comment by iaincollins** [comment]
To confirm, you don't call NextAuth.signin(), just post to the new end point.

You can do this using just a normal form post and/or in JavaScript.

You don't actually also need to do it in JavaScript unless you want to - a normal form post will work just fine, but if you want to you could have a sign in method like this in your sign in page (assumes email and password are stored in 'this.state').

````javascript
import React from 'react'
import Router from 'next/router'
import { NextAuth } from 'next-auth/client'

export default class extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      email: '',
      password: ''
    }
    this.handleSignin = this.handleSignin.bind(this)
  }

  handleSignin() {
    // Make sure we include the latest CSRF Token in the POST request as _csrf
    const formData = {
      _csrf: await NextAuth.csrfToken(),
      email: this.state.email,
      password: this.state.password
    }
  
    // Encoded form parser for sending data in the body
    const encodedForm = Object.keys(formData).map((key) => {
      return encodeURIComponent(key) + '=' + encodeURIComponent(formData[key])
    }).join('&')

    return fetch('/auth/password', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded'
      },
      body: encodedForm,
      credentials: 'same-origin'
    })
    .then(response => {
      if (response.ok) {
        Router.push(`/auth/callback}`)
      } else {
        Router.push(`/auth/error?action=signin&type=password&email=${this.state.email}`)
      }
    })
  }
  
}
````

If you find this works for you I am happy to look at both adding localStrategy support and at changing `NextAuth.signin()` to work with them (so it can be passed options to post to a local strategy, e.g. username or email, password, 2-factor-auth-token, etc).

None of these changes should be breaking - so anything you have done will keep working! - but should make it easier for anyone else in future.

**Comment by Jexah** [comment]
Hi, sorry to be a pain. I'm working on an application in which I require local authentication, with restricted login access. I haven't implemented it yet, but it looks like the above would satisfy the requirement for password authentication. My concern is regarding a case in which the client posts to `/auth/email/signin`, and has an account created for them via the default behavior. I don't want to maintain my own branch, and so I don't want to bastardize the source by ripping out and replacing behaviors. I attempted to override the route, but it seems like this is impossible without modifying the source.

My guess as to the best way to achieve this without branching the library is to define the `sendSignInEmail` function in `next-auth.functions.js` as an empty function. That way, at least the email won't send. Is this correct?

Aside: What is your guestimated timeline on built-in password authentication?

**Comment by iaincollins** [comment]
Hi @Jexah, not at all - requests for specific things are great.

I will try add the ability to define a localStrategy and not use email sign in route this weekend.

Thanks for sharing this! :-)

**Comment by Jexah** [comment]
Wow, that is fast! Thanks for the update! Quick question, unrelated and more out of curiosity than anything else, why don't you use native async/await instead of promises?

**Comment by kelleg1** [comment]
@Jexah I had the same question. Maybe it's that async/await are not as flexible? A talented longtime JavaScript programmer is probably also better versed with .then() style promise handling too. Personally, I would prefer async/await having a synchronous programming and C# background.

**Comment by iaincollins** [comment]
Hmm, that's a good question.

I suspect I probably had ideas around error handling with promises that have never been implemented, or other vague ideas about maybe executing other things concurrently while waiting for the response that were never followed up on.

Some of the updated client code was also gratefully received via a PR - though this may not actually be reflected in the git commits has had to quickly revert and then re-add the code in a preview version after it surfaced a bug that was since quashed.

The code existed in a starter project repo before it was forked off, and I think some it may predate async/await support in Next.js. The flip side of this is it means it's had time to iron out the kinks but some of it might do with refactoring.

Please do feel free to raise issues where you think the code could do with a review!

**Comment by kelleg1** [comment]
@Jexah Sounds cool. We probably ought not derail this topic from the conversation about local strategies, but something I am facing needing to do also is to use Mongoose with multiple connections. Since Mongoose uses the singleton pattern and only supports one connection per Mongoose instance, I have to use multiple instances if I want to use it. I do like it over the native driver.

I'd like to see your code. I was looking into Material UI too, but currently I use Reactstrap in addition to making my own components.

**Comment by Jexah** [comment]
Turns out I can't edit in mobile view, so tomorrow I'll merge this comment into the one above.

Do you accept pull requests on either repo? You can certainly run things concurrently with async await and (I think) the error handling is much nicer with regards to thread safety, though you need to make sure you catch everything, as you do with promises. 

An example of executing stuff while waiting for some response from async function.
```javascript
async func (){
//await timeout
}
...
async (req, res) => {
let a = func () // start processing async func
// start processing something else / do some sync stuff
...
// nothing left to do
res.send (await a); // either return or wait if still pending
}
```

Sorry for laziness and stuff, I'll edit when I get on my PC.

**Comment by iaincollins** [comment]
Hi, thanks for the feedback, agree it's best to move other discussions into another thread.

To confirm, I do know how await/async and Promises work. 

The issue was that await/async was not originally supported in older versions of node (for the server side code) and was not originally transpiled by babel in older versions of Next.js (for the client side portions of the code). This is no longer an issue in either case though.

If you use async/await do note they are not quite the same. Code in a function after an await statement **will not run until after the await call has returned** - this is unlike with a Promise, where code after a Promise in the same function will run without waiting for the Promise to return.
    
This is mostly not a problem in NextAuth but can change the behaviour of an app as when used it forces all code to execute serially, rather than allowing asynchronous calls to happen in Parallel.

I do always try to accept pull requests, but there aren't any unit tests for the auth yet so maybe best not to refactor any of the more sensitive stuff - specifically the client as it requires more testing, mostly to make sure Internet Explorer is happy, as it's more prone to having problems.

With regard to Mongoose, both the current version of Mongoose and the underlying Mongo driver use singletons (_object caching in require_) and connection pooling. The default for both is 5 connections and you can use the poolSize option in either to change it.

**Comment by Jexah** [comment]
I totally get the whole lack of support thing. With regards to the rest of your post: Functionally, async/await and Promises are *exactly* the same, but the implementation is just slightly different. I think while I was discussing functionality you thought I was talking about implementation, in which they are different. If you were discussing implementation, don't worry about reading the rest of this comment, but if you are saying that they functionally differ, please read the following.

While what you say regarding `await` is correct, the way you are implying it should be implemented is incorrect. Rarely in actual implementation do you do something like `await asyncFunc()`, as that does indeed block execution. Typically, you rarely -- if ever -- `await` a Promise until the value is absolutely required, as in the example I provided above. First I begin execution of an `async` function, and assigned the returned Promise to the `promise` variable, which is literally a native Promise. An `async` function is just a native wrapper for a function that returns a Promise, and `await` waits for resolution of the Promise. This means you can `await` a native Promise or `someAsyncFunction().then(res => { ... })`.

tl;dr `async` and `await` absolutely **do not force code to run serially**. They are just native syntactic sugar to make handling Promises nicer to read and write, and by extension, using them doesn't change the core underlying functionality of Promises, and thereby cannot have different functionality to that available in native Promises.

**Comment by iaincollins** [comment]
Hey @kelleg1, I've just published 1.8.0 to NPM which addresses this.

There is now the option to define a signIn() method in [next-auth.providers.js](https://github.com/iaincollins/next-auth/blob/master/example/next-auth.providers.js). It's easy to consume and there is an example of how to use it in the front end on [example/pages/auth/credentials.js](https://github.com/iaincollins/next-auth/blob/master/example/pages/auth/credentials.js).

If you define a signIn() method it enables POST requests to `/auth/signin`. You can pass any options you like to it by passing an object to `NextAuth.signin()`.

Additionally, if you want to disable email based sign in, just don't define a `sendSignInEmail()` function (or set it to null) and NextAuth will now longer set the routes for them (so people won't be able to create accounts unless you provide a specific endpoint outside of NextAuth for them to do that).

**Comment by tcash11** [comment]
@iaincollins You mention in your previous post that 

> There is now the option to define a signIn() method in next-auth.providers.js. 

Did you mean next-auth.functions.js? or is there something we need to add in the providers file as well for a username/pw type local strategy?

**Comment by iaincollins** [comment]
Sorry, yes thanks for the correction:

[next-auth.functions.js](https://github.com/iaincollins/next-auth/blob/master/example/next-auth.functions.js) and [auth/credentials.js](https://github.com/iaincollins/next-auth/blob/master/example/pages/auth/credentials.js) contain example usage.

I know it's not especially well documented, but I hope to improve it and the documentation for it.
